lib/src/sql/value/put.rs

```diff
@@ -26,6 +26,14 @@ impl Value {
 							v.insert(f.to_raw(), obj);
 						}
 					},
+					Part::Index(i) => match v.get_mut(&i.to_string()) {
+						Some(v) if v.is_some() => v.put(path.next(), val),
+						_ => {
+							let mut obj = Value::base();
+							obj.put(path.next(), val);
+							v.insert(i.to_string(), obj);
+						}
+					},
```

Comment:
It's worth noting that similar logic could be inserted in the following places:
<details>
<summary>Expand</summary>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/set.rs#L26-L47</li>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/del.rs#L26-L40</li>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/cut.rs#L10-L25</li>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/pick.rs#L10-L19C1</li>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/fetch.rs#L27-L39</li>
<li>https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/walk.rs#L14-L22</li>
</details>

Also, this line will probably re-interpret the path part as a field string instead of an index: https://github.com/surrealdb/surrealdb/blob/10f29fbb2670cb3461dd8383da851fb909c63027/lib/src/sql/value/diff.rs#L26

This possibility should be investigated prior to merging i.e. by converting numbers in strings to `Part::Index` instead of `Part::Field` or by shifting the burden of parsing numerical fields onto the parser (somehow).

Reply:
Yeah this logic needs to be implemented in all of those places. However, we can't shift this onto the parser though, as we never know what type we are currently working with (array or object).